version: "3.9"

services:
  traefik:
    image: traefik:v2.10
    command:
      - "--providers.docker=true"
      - "--providers.docker.exposedbydefault=false"
      - "--entrypoints.web.address=:80"
      - "--entrypoints.websecure.address=:443"
      - "--certificatesresolvers.myresolver.acme.httpchallenge=true"
      - "--certificatesresolvers.myresolver.acme.httpchallenge.entrypoint=web"
      - "--certificatesresolvers.myresolver.acme.email=${TRAEFIK_LE_EMAIL}"
      - "--certificatesresolvers.myresolver.acme.storage=/letsencrypt/acme.json"
      - "--log.level=DEBUG"
      - "--entryPoints.metrics.address=:8082"
      - "--metrics.prometheus=true"
      - "--metrics.prometheus.entryPoint=metrics"
      - "--metrics.prometheus.buckets=0.1,0.3,1.2,5.0"
    ports:
      - 80:80
      - 443:443
    volumes:
      - letsencrypt:/letsencrypt
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
    restart: on-failure
    logging: &logging
      driver: "json-file"
      options:
        max-size: "200k"
        max-file: "10"
    labels:
      - "traefik.enable=true"
      - "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https"
      - "traefik.http.routers.http-catchall.rule=hostregexp(`{host:.+}`)"
      - "traefik.http.routers.http-catchall.entrypoints=web"
      - "traefik.http.routers.http-catchall.middlewares=redirect-to-https"

  prometheus:
    image: prom/prometheus:v2.47.1
    volumes:
      - prometheus-data:/prometheus
      - /home/ubuntu/config/prometheus.yml:/etc/prometheus/prometheus.yml
    extra_hosts:
      - "host.docker.internal:host-gateway"
    restart: on-failure
    logging: *logging

  grafana:
    image: grafana/grafana:10.0.8
    volumes:
      - grafana-data:/var/lib/grafana
      - /home/ubuntu/config/grafana-provisioning:/etc/grafana/provisioning
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=o4MkVu0elSCFf4dD
      - GF_USERS_ALLOW_SIGN_UP=false
      - GF_INSTALL_PLUGINS=grafana-piechart-panel
      - GF_SERVER_ROOT_URL=https://${SUBDOMAIN:-$ENVIRONMENT-$CARDANO_NETWORK}.govtool.byron.network/grafana
      - GF_SERVER_SERVE_FROM_SUB_PATH=true
    restart: on-failure
    logging: *logging
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.grafana.rule=Host(`${SUBDOMAIN:-$ENVIRONMENT-$CARDANO_NETWORK}.govtool.byron.network`) && PathPrefix(`/grafana`)"
      - "traefik.http.routers.grafana.entrypoints=websecure"
      - "traefik.http.routers.grafana.tls.certresolver=myresolver"
      - "traefik.http.services.grafana.loadbalancer.server.port=3000"

  postgres:
    image: postgres:11.18-alpine
    environment:
      - POSTGRES_LOGGING=true
      - POSTGRES_DB_FILE=/run/secrets/postgres_db
      - POSTGRES_PASSWORD_FILE=/run/secrets/postgres_password
      - POSTGRES_USER_FILE=/run/secrets/postgres_user
    secrets:
      - postgres_password
      - postgres_user
      - postgres_db
    volumes:
      - postgres:/var/lib/postgresql/data
    restart: on-failure
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 10s
      timeout: 5s
      retries: 5
    command: ${POSTGRES_ARGS:--c maintenance_work_mem=1GB -c max_parallel_maintenance_workers=4}
    logging: *logging

  cardano-node:
    image: inputoutput/cardano-node:8.0.0
    entrypoint:
      - "cardano-node"
      - "run"
      - "--topology"
      - "/configuration/topology.json"
      - "--database-path"
      - "/data/db"
      - "--socket-path"
      - "/ipc/node.socket"
      - "--host-addr"
      - "0.0.0.0"
      - "--port"
      - "3001"
      - "--config"
      - "/configuration/config.json"
      - "+RTS"
      - "-N2"
      - "-I0"
      - "-A16m"
      - "-qg"
      - "-qb"
      - "--disable-delayed-os-memory-return"
      - "-RTS"
    environment:
      - NETWORK=${CARDANO_NETWORK:-preprod}
    volumes:
      - node-db:/data/db
      - node-ipc:/ipc
      - /home/ubuntu/config/cardano-node:/configuration
    restart: on-failure
    healthcheck:
      test: ["CMD-SHELL", "curl -f 127.0.0.1:12788 || exit 1"]
      interval: 60s
      timeout: 10s
      retries: 5
    logging: *logging

  cardano-db-sync:
    image: inputoutput/cardano-db-sync:13.1.1.2-docker
    environment:
      - NETWORK=${CARDANO_NETWORK:-preprod}
      - POSTGRES_HOST=postgres
      - POSTGRES_PORT=5432
      - RESTORE_SNAPSHOT=${RESTORE_SNAPSHOT:-}
      - RESTORE_RECREATE_DB=N
      - EXTRA_DB_SYNC_ARGS=${EXTRA_DB_SYNC_ARGS:-}
    depends_on:
      cardano-node:
        condition: service_healthy
      postgres:
        condition: service_healthy
    secrets:
      - postgres_password
      - postgres_user
      - postgres_db
    volumes:
      - db-sync-data:/var/lib/cexplorer
      - node-ipc:/node-ipc
    restart: on-failure
    logging: *logging

  postgres-fakedbsync:
    image: postgres:15-alpine
    environment:
      - POSTGRES_LOGGING=true
      - POSTGRES_DB=vva
      - POSTGRES_PASSWORD=test
      - POSTGRES_USER=postgres
    volumes:
      - postgres-fakedbsync:/var/lib/postgresql/data
      - /home/ubuntu/config/fakedbsync_init.d:/docker-entrypoint-initdb.d
    restart: on-failure
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 10s
      timeout: 5s
      retries: 5
    command: ${POSTGRES_ARGS:--c maintenance_work_mem=1GB -c max_parallel_maintenance_workers=4}
    logging: *logging

  vva-be:
    image: ${VVA_BE_REPO:-733019650473.dkr.ecr.eu-west-1.amazonaws.com/backend}:${TAG:-main-dev-preprod}
    command: /usr/local/bin/vva-be -c /run/secrets/vva-be-config.json start-app
    depends_on:
      cardano-node:
        condition: service_healthy
      postgres:
        condition: service_healthy
      postgres-fakedbsync:
        condition: service_healthy
    secrets:
      - vva-be-config.json
    restart: on-failure
    logging: *logging
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.vva-be.rule=Host(`${SUBDOMAIN:-$ENVIRONMENT-$CARDANO_NETWORK}.govtool.byron.network`) && PathPrefix(`/api`)"
      - "traefik.http.middlewares.vva-be-stripprefix.stripprefix.prefixes=/api"
      - "traefik.http.routers.vva-be.middlewares=vva-be-stripprefix@docker"
      - "traefik.http.routers.vva-be.entrypoints=websecure"
      - "traefik.http.routers.vva-be.tls.certresolver=myresolver"
      - "traefik.http.services.vva-be.loadbalancer.server.port=9876"

  vva-fe:
    image: ${VVA_FE_REPO:-733019650473.dkr.ecr.eu-west-1.amazonaws.com/frontend}:${TAG:-main-dev-preprod}
    depends_on:
      cardano-node:
        condition: service_healthy
      postgres:
        condition: service_healthy
    restart: on-failure
    logging: *logging
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.vva-fe.rule=Host(`${SUBDOMAIN:-$ENVIRONMENT-$CARDANO_NETWORK}.govtool.byron.network`)"
      - "traefik.http.routers.vva-fe.entrypoints=websecure"
      - "traefik.http.routers.vva-fe.tls.certresolver=myresolver"
      - "traefik.http.services.vva-fe.loadbalancer.server.port=80"

secrets:
  postgres_db:
    file: /home/ubuntu/config/dbsync-secrets/postgres_db
  postgres_password:
    file: /home/ubuntu/config/dbsync-secrets/postgres_password
  postgres_user:
    file: /home/ubuntu/config/dbsync-secrets/postgres_user
  vva-be-config.json:
    file: /home/ubuntu/config/vva-be-config.json

volumes:
  letsencrypt:
  db-sync-data:
  grafana-data:
  postgres:
  postgres-fakedbsync:
  prometheus-data:
  node-db:
  node-ipc:
